src/comm/aca_message_pulsar_consumer.cpp

```diff
@@ -0,0 +1,142 @@
+// Copyright 2019 The Alcor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "aca_message_pulsar_consumer.h"
+#include "goalstate.pb.h"
+#include "aca_comm_mgr.h"
+#include "aca_log.h"
+
+using aca_comm_manager::Aca_Comm_Manager;
+using pulsar::Client;
+using pulsar::ClientConfiguration;
+using pulsar::ConsumerConfiguration;
+using pulsar::Consumer;
+using pulsar::Message;
+using pulsar::Result;
+
+
+namespace messagemanager
+{
+MessageConsumer::MessageConsumer(string brokers, string subscription_name)
+{
+  setBrokers(brokers);
+  setSubscriptionName(subscription_name);
+
+  // TO DO: Construct the configuration
+
+  ACA_LOG_DEBUG("Broker list: %s\n", this->brokers_list.c_str());
+  ACA_LOG_DEBUG("Consumer subscription name: %s\n", this->subscription_name.c_str());
+
+  // Create the consumer and client
+  this->ptr_consumer = new Consumer();
+  this->ptr_client= new Client(brokers,this->client_config);
+}
+
+MessageConsumer::~MessageConsumer()
+{
+  //delete ptr_consumer;
+  delete ptr_client;
+}
+
+string MessageConsumer::getBrokers() const
+{
+  return this->brokers_list;
+}
+
+string MessageConsumer::getLastTopicName() const
+{
+  return this->topic_name;
+}
+
+string MessageConsumer::getSubscriptionName() const
+{
+  return this->subscription_name;
+}
+
+void MessageConsumer::setSubscriptionName(string subscription_name)
+{
+  this->subscription_name = subscription_name;
+}
+
+bool MessageConsumer::consumeDispatched(string topic)
```

Comment:
You will need to update aca_main.cpp under the src directory to:
1. line 33-35: update the default from kafka to pulsar
2. line 46-48: rename the global variable to pulsar name
3. line 151-153: update help text to pulsar
4. line 191: call your new MessagePulsarConsumer class instead

Reply:
Ok, and one question: I don't clearly figure out what's the function of dispatcher in kafka consumer(it seems to be a loop function), so I simply replace it with a few if/else. Is there anything I need to do?

Reply:
Since the kafka consumer is not being used, the plan is to switch over all the kafka related entry points to use your new MessagePulsarConsumer class by changing the lines of code mentioned above. Therefore, you don't need to worry about the dispatcher in kafka consumer since it will not be used.